### DISCUSSION TITLE
Improve persistence mechanism

### DISCUSSION THREAD

**Discussion by laellyo** [cloud-carbon-footprint/cloud-carbon-footprint#1144] [Feature suggestions]
Hello,

Thanks to your documentation, I have easily deployed a CCF instance for my company in a AKS cluster, with a mongodb database as persistence component, and connect it to Azure to retrieve consumption data.
As we are a quite big company, the number of assets is huge at worldwide level, causing the CCF API pod to restart due to exceeded memory outage. An example of error :

```bash
2023-05-17T10:18:56.615Z [Cache] info: Setting new estimates to cache... 

<--- Last few GCs --->

[29:0x7f155c6c83f0] 167220347 ms: Mark-sweep 914.8 (1044.0) -> 907.5 (1042.5) MB, 1432.3 / 0.0 ms  (average mu = 0.129, current mu = 0.056) allocation failure scavenge might not succeed
[29:0x7f155c6c83f0] 167221776 ms: Mark-sweep 916.5 (1042.5) -> 911.9 (1041.0) MB, 1399.0 / 0.0 ms  (average mu = 0.078, current mu = 0.021) allocation failure scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
error Command failed with signal "SIGABRT".
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

I tried to increase the memory, but definitively, it is not a good option, as the amount of data to process is certainly too high to be fully loaded in memory.
By using a MongoDB database, I was expecting that CCF streams the data to MongoDB continously, but it seems that it pushes it only when finishing the full request process. (all the Azure resources groups for a given time interval). 
By reducing the time interval to one day thanks to a custom request like `https://x.x.x.x/api/footprint?start=2023-01-01&end=2023-01-02&groupBy=day&limit=500`. CCF API is able to retrieve all data and push them to MongoDB. Indeed, I had the confirmation that a mongo db has been created and contains data, as proved by this request:

```bash
rs0 [direct: primary] test> show dbs
admin    80.00 KiB
ccf      50.98 MiB # My Cloud Carbon Footprint database!
config  200.00 KiB
local    52.39 MiB
```

However, if I try to increase the time range to one week or one month, the CCF API pod is restarting due to memory issue without persisting collected data in MongoDB.

Could you confirm that you have a similar behavior when collecting a lot of data? Or does it come from a misconfiguration?
If this behavior is "by design", it would be great to change it to stream collected data to MongoDB to avoid heap out of memory and increase platform reliability. 
Ideally, I would like to be able to request data for a full year, and let the API paginate, collect, process and store metrics into the database.

**Comment by 4upz** [comment]
Hi @laellyo thanks for opening this discussion! We can confirm that this is expected behavior when making requests that potentially contain a large amount of data. We detail some findings and potential solutions within our documentation for [Performance Considerations](https://www.cloudcarbonfootprint.org/docs/performance-considerations#improving-query-performance-and-handling-large-data).

My suggestion is to check out our script for [seeding data](https://www.cloudcarbonfootprint.org/docs/data-persistence-and-caching#seeding-cache-file) into your database. This CLI command follows similar behavior that you are describing in which it allows you to choose a "split" fetch method in which it streams that data into your database in smaller chunks for a given date range. This is our recommended approach for handling the backfilling of large historical data. 

Unfortunately, we don't have prioritized plans for bringing this behavior into the main app's API at the moment, but are open to PR contributions or feedback if this issue is recognized as a common and high priority fix for the community.
